pshtt.py

```diff
@@ -349,38 +369,39 @@ def canonical_endpoint(http, httpwww, https, httpswww):
     # or like:
     #   https:// -> 200, http:// -> http://www
 
+    at_least_one_www_used = httpswww.live or httpwww.live
+
+    def root_unused(endpoint):
+        return (
+            endpoint.redirect or
+            (not endpoint.live) or
+            endpoint.https_bad_hostname or  # harmless for http endpoints
+            (not str(endpoint.status).startswith("2"))
+        )
+
+    def root_down(endpoint):
+        return (
+            (not endpoint.live) or
+            endpoint.https_bad_hostname or
+            (not str(endpoint.status).startswith("2"))
+        )
+
+    def goes_to_www(endpoint):
+        return (
+            endpoint.redirect_immediately_to_www and
+            (not endpoint.redirect_immediately_to_external)
+        )
+
+    all_roots_unused = root_unused(https) and root_unused(http)
+
+    all_roots_down = root_down(https) and root_down(http)
+
     is_www = (
-        (httpswww.live or httpwww.live) and (
-            (
-                https.redirect or
-                (not https.live) or
-                https.https_bad_hostname or
-                (not str(https.status).startswith("2"))
-            ) and (
-                http.redirect or
-                (not http.live) or
-                (not str(http.status).startswith("2"))
-            )
-        ) and (
-            (
-                (
-                    (not https.live) or
-                    https.https_bad_hostname or
-                    (not str(https.status).startswith("2"))
-                ) and
-                (
-                    (not http.live) or
-                    (not str(http.status).startswith("2"))
-                )
-            ) or
-            (
-                https.redirect_immediately_to_www and
-                (not https.redirect_immediately_to_external)
-            ) or
-            (
-                http.redirect_immediately_to_www and
-                (not http.redirect_immediately_to_external)
-            )
+        at_least_one_www_used and
+        all_roots_unused and (
+            all_roots_down or
+            goes_to_www(https) or
+            goes_to_www(http)
         )
```

Comment:
The above refactoring of the `is_www` block should be identical to the version before, just easier to read and ebug.